src/optimisation/Optimisation.jl

```diff
@@ -1,324 +1,327 @@
-    p .= vi[:]
-
-    # If linking mutated, we need to link once more.
-    linked && DynamicPPL.link!!(vi, model)
-
-    return p
-end
+"""
+    OptimizationProblem(me_prob::ModeEstimationProblem)
 
-function transform!!(p::AbstractArray, vi::DynamicPPL.VarInfo, model::DynamicPPL.Model, ::constrained_space{false})
-    linked = DynamicPPL.istrans(vi)
-    if linked
-        vi = DynamicPPL.invlink!!(vi, model)
+Create a SciML `OptimizationProblem` from a `ModeEstimationProblem`.
+"""
+function OptimizationProblem(me_prob::ModeEstimationProblem)
+    c = me_prob.constraints
+    # Note that OptimLogDensity is a callable that evaluates the model with given
+    # parameters. Hence we can use it as the objective function.
+    f = OptimizationFunction((x, _) -> me_prob.log_density(x), me_prob.adtype; cons=c.cons)
+    if !has_constraints(me_prob)
+        opt_prob = OptimizationProblem(f, me_prob.init_value)
+    else
+        opt_prob = OptimizationProblem(
+            f, me_prob.init_value;
+            lcons=c.lcons, ucons=c.ucons, lb=c.lb, ub=c.ub
+        )
     end
-    vi = DynamicPPL.unflatten(vi, p)
-    vi = DynamicPPL.link!!(vi, model)
-    p .= vi[:]
-
-    # If linking mutated, we need to link once more.
-    !linked && DynamicPPL.invlink!!(vi, model)
-
-    return p
+    return opt_prob
 end
 
-function transform(p::AbstractArray, vi::DynamicPPL.VarInfo, model::DynamicPPL.Model, con::constrained_space)
-    return transform!!(copy(p), vi, model, con)
-end
-
-abstract type AbstractTransform end
-
-struct ParameterTransform{T<:DynamicPPL.VarInfo,M<:DynamicPPL.Model, S<:constrained_space} <: AbstractTransform
-    vi::T
-    model::M
-    space::S
-end
+"""
+    ModeResult(prob::ModeEstimationProblem, solution::AbstractVector)
 
-struct Init{T<:DynamicPPL.VarInfo,M<:DynamicPPL.Model, S<:constrained_space} <: AbstractTransform
-    vi::T
-    model::M
-    space::S
-end
+Create a `ModeResult` for a given estimation problem and a solution given by `solve`.
 
-function (t::AbstractTransform)(p::AbstractArray)
-    return transform(p, t.vi, t.model, t.space)
+`Optimization.solve` returns its own result type. This function converts that into the
+richer format of `ModeResult`. It also takes care of transforming them back to the original
+parameter space in case the optimization was done in a transformed space.
+"""
+function ModeResult(prob::ModeEstimationProblem, solution::AbstractVector)
+    solution_values = solution.u
+    ld = prob.log_density
+    if prob.linked
+        ld = Accessors.@set ld.varinfo = DynamicPPL.unflatten(ld.varinfo, solution_values)
+        ld = Accessors.@set ld.varinfo = DynamicPPL.invlink(ld.varinfo, ld.model)
+        solution_values = ld.varinfo[:]
+    end
+    # Store the parameters and their names in a NamedArray.
+    varnames = variable_names(prob.log_density)
+    vmat = NamedArray(solution_values, varnames)
+    return ModeResult(vmat, solution, -solution.objective, prob.log_density)
```

Comment:
The following should do the trick:

```julia
    varinfo_new = DynamicPPL.unflatten(varinfo, solution.u)
    # `getparams` performs the invlinking, if needed, etc.
    vns_vals_iter = getparams(prob.model, varinfo_new)
    syms = map(Symbol ∘ first, vns_vals_iter)
    vals = map(last, vns_vals_iter)
    return ModeResult(
        NamedArray(vals, syms),
        solution,
        -solution.objective,
        prob.log_density
    )
```

Though probably need to fix some of the references.

Reply:
Nice, done.